test/aleph/http/multipart_test.clj

```diff
@@ -188,39 +189,73 @@
 
     (.close ^java.io.Closeable s)))
 
-(defn- pack-chunk [{:keys [content] :as chunk}]
-  (cond-> (dissoc chunk :file)
-    (not (string? content))
-    (dissoc :content)))
-
-(defn- decode-handler [req]
-  (let [chunks (-> req
-                   mp/decode-request
-                   s/stream->seq)]
-    {:status 200
-     :body (pr-str (map pack-chunk chunks))}))
-
-(defn- test-decoder [port url raw-stream?]
-  (let [s (http/start-server decode-handler {:port port
-                                             :raw-stream? raw-stream?})
-        chunks (-> (http/post url {:multipart parts})
-                   (deref 1e3 {:body "timeout"})
-                   :body
-                   bs/to-string
-                   clojure.edn/read-string
-                   vec)]
-    (is (= 6 (count chunks)))
-
-    ;; part-names
-    (is (= (map :part-name parts)
-           (map :part-name chunks)))
-
-    ;; content
-    (is (= "CONTENT1" (get-in chunks [0 :content])))
-
-    ;; mime type
-    (is (= "text/plain" (get-in chunks [2 :mime-type])))
-    (is (= "application/png" (get-in chunks [3 :mime-type])))
+(defn- coerce-chunk-content [{:keys [content file] :as chunk}]
+  (let [content' (cond
+                   (string? content)
+                   content
+
+                   (some? file)
+                   (slurp (.getAbsolutePath ^java.io.File file))
+
+                   :else
+                   (bs/to-string content))]
+    (assoc chunk :content content')))
+
+(defn- pack-chunk [chunk]
+  (-> chunk
+      (coerce-chunk-content)
+      (dissoc :file :raw-http-data)))
+
+(defn- decode-handler [req options]
+  (let [chunks (mp/decode-request req options)]
+    (->
+     (d/loop [body []]
+       (d/chain' (s/take! chunks)
+                 (fn [{:keys [file] :as chunk}]
+                   (Thread/sleep 20)
```

Comment:
Why sleep? Will user handlers have to do this, too?

Reply:
It's basically how I demonstrated the proposed implementation wasn't working as expected.

`stream/on-closed` is called when the data has been fully read from the `body` and it has been closed.
However, nothing guarantees the data has been taken on the consumer side. Some parts might deleted
before the client has the time to consume them.

Without that `Thread/sleep` which simulates time spent handling a part, all the tests pass.

Oleksii also gives a bit of context around it [1]

> At all. chunks stream will be eventually closed by the implementation as soon as decoded parts are exhausted. on-close is implemented there to release all resources that we didn't consume from the decoder yet. But the stream is only a delivery mechanism, so when it's closed we know that values were conveyed to a consumer, but we don't know if they are still in use. You can pass them to any async processing pipeline you want, and the closing of the stream means nothing for us.

[1] : https://github.com/clj-commons/aleph/pull/432#issuecomment-449376344

Reply:
OK